components/batch-system/src/batch.rs

```diff
@@ -288,21 +306,31 @@ impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
             if !batch.normals.is_empty() {
                 for (i, p) in batch.normals.iter_mut().enumerate() {
                     let len = self.handler.handle_normal(p);
+                    batch.counters[i] += 1;
                     if p.is_stopped() {
-                        exhausted_fsms.push((i, None));
-                    } else if len.is_some() {
-                        exhausted_fsms.push((i, len));
+                        reschedule_fsms.push((i, ReschedulePolicy::Remove));
+                    } else {
+                        if batch.counters[i] > 3 {
+                            hot_fsm_count += 1;
+                            if hot_fsm_count & 1 == 1 {
```

Comment:
Does it mean to just re-schedule a half of hot FSMs? Shall we add some comments here?